Mrcs monitor metrics

CCatania01 opened the pull request:
CMRI Menu item re-order
SerialMon Updates with node and packet filtering
<dle> added to Raw data packet display
CMRInetMetrics tests added

CCatania01 commented:
What is the best method using Netbeans to resolve the conflicts?  I assume I need to refresh my local repository from the master.

silverailscolo commented:
See [JMRI NetBeans Help](http://jmri.org/help/en/html/doc/Technical/NetBeans.shtml) under **Updating the code from Git**
In the code you will see the overlapping edits marked with blocks of ``>>>>`` and ``<<<<`` for HEAD (your work) and master (someone else's recent contribs).

CCatania01 commented:
Do I do a branch switch to "master" or "origin/master" for the Pull for Upstream?

CCatania01 commented:
Closing this PR and starting over.
Chuck Catania

bobjacobsen commented:
I'm going to reopen this as WIP, since that makes it much easier to see what's going wrong.

bobjacobsen commented:
@CCatania01 Right now, I'm not seeing any conflicts. Maybe something was merged?

CCatania01 commented:
The initial file conflict errors were strange and did not reflect my version of the files.
I have reset my environment from a full backup done just prior to attempting to submit the PR.  I checked the files and changes from NetBeans and they are correct. I did a clean build in NetBeans and ran my dual-connection regression test.  All functions are what I expected from earlier unit testing.

Rather than trying to understand my GitHub procedural errors, I would prefer to start from this known point. That is why I closed the PR.

Chuck

On Sep 1, 2017, at 8:09 AM, Bob Jacobsen wrote:

> @CCatania01 Right now, I'm not seeing any conflicts. Maybe something was merged?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

CCatania01 commented:
Now what?  Cannot do Push to Upstream.

bobjacobsen commented:
@CCatania01 
> Now what? Cannot do Push to Upstream.

Try pulling from your own (CCatania01) repository. "Pull" will fetch from your repository, then merge it in.  Apparently  there are differences there that can't be automatically merged, so you might get some conflicts that you'll have to resolve:  edit, git add, git commit.  Then push back.

CCatania01 commented:
The commit seemed to work, but no the push.

==[IDE]== Sep 2, 2017 7:07:02 PM Committing...
Git Commit
----------
git add /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/CMRIMenu.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsCollector.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsData.java /Users/admin/mrcs-cmri/2017-github/JMRI /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetMetricsAction.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/SerialTrafficController.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetManagerFrame.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetMetricsFrame.java
git commit -m CMRI menu item reordered
SerialMon display updates
CMRI message filtering by type and node number
Added <dle> to raw monitor data output
Added CMRInetMetrics tests /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/CMRIMenu.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsCollector.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsData.java /Users/admin/mrcs-cmri/2017-github/JMRI /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetMetricsAction.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/SerialTrafficController.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetManagerFrame.java /Users/admin/mrcs-cmri/2017-github/JMRI/java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetMetricsFrame.java
Commit Log
revision   : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
author     : Chuck Catania <cpcrr@charter.net>
date       : Sep 2, 2017 7:07:34 PM
summary    : CMRI menu item reordered
             SerialMon display updates
             CMRI message filtering by type and node number
             Added <dle> to raw monitor data output
             Added CMRInetMetrics tests

INFO: End of Commit

==[IDE]== Sep 2, 2017 7:07:47 PM Committing... finished.
==[IDE]== Sep 2, 2017 7:08:43 PM Pushing - JMRI
git branch
git submodule status
git push https://github.com/CCatania01/JMRI.git refs/heads/MRCS-monitor-metrics:refs/heads/MRCS-monitor-metrics

Remote Repository Updates
Branch Update : MRCS-monitor-metrics
Old Id        : df7c7b1b2fee784e6dabd8635c588cd80e5d6e4a
New Id        : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
Result        : REJECTED_NONFASTFORWARD

Local Repository Updates
==[IDE]== Sep 2, 2017 7:08:58 PM Pushing - JMRI finished.
==[IDE]== Sep 2, 2017 7:10:00 PM Pushing - JMRI
git branch
git remote -v
setting up remote: origin
git submodule status
git push https://github.com/CCatania01/JMRI.git refs/heads/MRCS-monitor-metrics:refs/heads/MRCS-monitor-metrics

Remote Repository Updates
Branch Update : MRCS-monitor-metrics
Old Id        : df7c7b1b2fee784e6dabd8635c588cd80e5d6e4a
New Id        : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
Result        : REJECTED_NONFASTFORWARD

Local Repository Updates
==[IDE]== Sep 2, 2017 7:10:06 PM Pushing - JMRI finished.
==[IDE]== Sep 2, 2017 7:10:20 PM Pushing - JMRI
git branch
git submodule status
git push https://github.com/CCatania01/JMRI.git refs/heads/MRCS-monitor-metrics:refs/heads/MRCS-monitor-metrics

Remote Repository Updates
Branch Update : MRCS-monitor-metrics
Old Id        : df7c7b1b2fee784e6dabd8635c588cd80e5d6e4a
New Id        : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
Result        : REJECTED_NONFASTFORWARD

Local Repository Updates
==[IDE]== Sep 2, 2017 7:10:26 PM Pushing - JMRI finished.
==[IDE]== Sep 2, 2017 7:11:24 PM Pushing - JMRI
git branch
git submodule status
git push https://github.com/CCatania01/JMRI.git refs/heads/MRCS-monitor-metrics:refs/heads/MRCS-monitor-metrics

Remote Repository Updates
Branch Update : MRCS-monitor-metrics
Old Id        : df7c7b1b2fee784e6dabd8635c588cd80e5d6e4a
New Id        : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
Result        : REJECTED_NONFASTFORWARD

Local Repository Updates
==[IDE]== Sep 2, 2017 7:11:33 PM Pushing - JMRI finished.

bobjacobsen commented:
```
Remote Repository Updates
Branch Update : MRCS-monitor-metrics
Old Id : df7c7b1
New Id : 30e1fbc1de94c5118e8ed12ce47c59998b4b7ced
Result : REJECTED_NONFASTFORWARD
```

Right. So you _need_ _to_ _pull_ _back_ _that_ _repository_ _into_ _your_ _local_ _one_.

See prior comment:
> Try pulling from your own (CCatania01) repository. "Pull" will fetch from your repository, 
> then merge it in. Apparently there are differences there that can't be automatically 
> merged, so you might get some conflicts that you'll have to resolve: 
> edit, git add, git commit. Then push back.

CCatania01 commented:
I am tyring to recover from a bad error on my part. I think I am back to where I was before the error.
I am getting this error when I open a panel with two connections defined.  A single connection seems to work. 

Any clue as to what is broken?  If some Git commands can get this straightened out, please  advise as to what they should be.

2017-09-14 17:20:30,880 util.Log4JUtil                        INFO  - ****** JMRI log ******* [main]
2017-09-14 17:20:33,340 util.Log4JUtil                        INFO  - This log is appended to file: /Users/Development/Library/Preferences/JMRI/log/messages.log [main]
2017-09-14 17:20:33,341 util.Log4JUtil                        INFO  - This log is stored in file: /Users/Development/Library/Preferences/JMRI/log/session.log [main]
2017-09-14 17:20:33,351 apps.Apps                             INFO  - PanelPro version 4.9.4ish+Development+20170915T0020Z+Ra5545fe starts under Java 1.8.0_144 on Mac OS X x86_64 v10.9.5 at Thu Sep 14 17:20:33 PDT 2017 [main]
2017-09-14 17:20:38,366 apps.Apps                             INFO  - Starting with profile Dual_CMRI_Connections.3f0749d2 [main]
2017-09-14 17:20:38,699 node.NodeIdentity                     INFO  - Using jmri-D49A20066074-3f0749d2 as the JMRI Node identity [main]
2017-09-14 17:20:39,243 serialdriver.SerialDriverAdapter      INFO  - cu.usbserial-A101OD7U port opened at 28800 baud with DTR: true RTS: true DSR: true CTS: true  CD: true [main]
2017-09-14 17:20:39,262 serial.SerialTrafficController        INFO  - C2-SerialTrafficController ctor called [main]
2017-09-14 17:20:39,402 serialdriver.SerialDriverAdapter      INFO  - cu.usbserial-A101O9WE port opened at 28800 baud with DTR: true RTS: true DSR: true CTS: true  CD: true [main]
2017-09-14 17:20:39,405 serial.SerialTrafficController        INFO  - C2-SerialTrafficController ctor called [main]
2017-09-14 17:20:40,424 util.FileUtilSupport                  INFO  - File path program: is /Users/Development/mrcs-cmri/2017-github/JMRI/ [main]
2017-09-14 17:20:40,426 util.FileUtilSupport                  INFO  - File path preference: is /Users/Development/Library/Preferences/JMRI/Dual_CMRI_Connections/ [main]
2017-09-14 17:20:40,427 util.FileUtilSupport                  INFO  - File path profile: is /Users/Development/Library/Preferences/JMRI/Dual_CMRI_Connections/ [main]
2017-09-14 17:20:40,428 util.FileUtilSupport                  INFO  - File path settings: is /Users/Development/Library/Preferences/JMRI/ [main]
2017-09-14 17:20:40,429 util.FileUtilSupport                  INFO  - File path home: is /Users/Development/ [main]
2017-09-14 17:20:40,429 util.FileUtilSupport                  INFO  - File path scripts: is /Users/Development/mrcs-cmri/2017-github/JMRI/jython/ [main]
2017-09-14 17:20:49,782 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-14 17:20:49,783 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-14 17:20:49,789 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]

devel-bobm commented:
@CCatania01 , your problem with multiple C/MRI connections is likely related to a comment in #2953 by  @bobjacobsen on 15Apr2017.

CCatania01 commented:
Multiple C/MRI connections are working in 4.9.3ish with the updated CMRI changes.  As you are aware, I may have corrupted my local environment when creating the monitor-metrics branch.  
I believe I did recover back to the point of my last push upstream for 4.9.3. Creating the monitor-metrics branch is where the observed problem started.  I am not confident that the branch has all of the proper master files.
Which master do I use for a pull "master" or "origin/master".  Once I am back in sync, I can re-enter the changes for the monitor and metrics.

bobjacobsen commented:
@CCatania01:  You need to break the habit of "recovering" from backups, older copies, etc.  That's neither useful nor appropriate for a Git environment.

The best thing you can do now is to throw away your local content (the cloned repository on your local disk), resolve to never use any backup of the old content (my Mac is configured to exclude git directories from backups, because (1) it's not needed and (2) that saves a bunch of backup space), and recreate a local cloned repository from scratch.  

I don't know how to do the following steps in NetBeans.  Perhaps somebody else can comment on that.  From the command line, you recreate a local cloned repository from scratch with:

```
git clone https://github.com/JMRI/JMRI.git 
git remote set-url --push origin https://github.com/CCatania01/JMRI.git
git remote add CCatania01 https://github.com/CCatania01/JMRI.git
git fetch CCatania01
```

That's created a new clone of our main GitHub repository, configured the connections to your existing GitHub repository, and fetched anything different in your own GitHub repository. That's all clean content.

Now, finally, get back the [MRCS-monitor-metrics branch](https://github.com/CCatania01/JMRI/tree/MRCS-monitor-metrics) from your own GitHub repository:

```
git branch --track MRCS-monitor-metrics CCatania01/MRCS-monitor-metrics
```

What you'll then have in the current directory is a `MRCS-monitor-metrics` branch with exactly the content that this PR has, all connected up properly to GitHub.  (Please check that; check that the changes are what you see in the [online diffs](https://github.com/JMRI/JMRI/compare/master...CCatania01:MRCS-monitor-metrics), and do a `git status` to make sure that the branch name is correct)

From there, you can hook it up to NetBeans, make any changes necessary, commit them, and push them back to your GitHub repository.  That'll include them in this PR automatically.

One necessary change is to fix the [conflict you can see above](https://github.com/JMRI/JMRI/pull/4035/conflicts).  After figuring out how to do that, you can do that directly (by editing in a change that'll fix it, commit, push to your repository: If the file contents are consistent, there's no conflict).  Or you can pull in the current JMRI master content (something you may want to do anyway) and resolve it with:

```
git pull master
```
You'll see messages about the particular file(s) with the conflict(s) - edit those, then tell git they're reading to go with e.g.
```
git add java/src/jmri/jmrix/cmri/serial/serialmon/SerialFilterFrame.java
```
i.e. do a `git add` for each resolved file, then
```
git commit
```
with no other arguments. It'll show you what you're committing as part of the final merge, ask you to enter a comment (or accept it's default one), save the message file, and then the merge is done.

Bottom line: Git directories should be managed with git.  If you do that, you shouldn't get into trouble.  If you do get into trouble, _git_ has tools of more and more power/complexity that can help resolve them, but if that doesn't work, the only other approach is to go back to the most recent correct content that Git has.  Other methods of recovery (a) are unlikely to work reliably and (b) waste a lot of time as they make things worse.

CCatania01 commented:
Thank you.  Yes, I have learned my lesson(s) the hard way. 
Earlier, re-starting from a known point was what I wanted to do by closing this PR.
I appreciate all of the strong advice given over the last couple weeks.

Do I close this PR now?
Thanks,
Chuck Catania

CCatania01 commented:
Branch is up to date.  Two commits done.
Last login: Fri Sep 15 16:21:54 on ttys000
Chucks-MacBook-Pro:~ Development$ cd /Users/Development/mrcs-cmri/2017-github/JMRI
Chucks-MacBook-Pro:JMRI Development$ git status
On branch MRCS-monitor-metrics
Your branch is up-to-date with 'CCatania01/MRCS-monitor-metrics'.

nothing to commit, working directory clean
Chucks-MacBook-Pro:JMRI Development$ 

The following files show an error tag in Netbeans.  I cannot do a clean compile.
This is the point two weeks ago where things started to come unraveled.  What is the process for moving forward?

jmri.jmrit.vsdecoder.listener.ListeningSpot.java
jmri.jmrit.vsdecoder.swing.ManageLocationsAction
jmri.util.PhysicalLocation.java
jmri.util.PhysicalLocation.PhysicalLocationPanel

rhwood commented:
If the files in NetBeans show something like this:
```
<<<<<<< MRCS-monitor-metrics
    SerialFilterFrame curFrame;   
    
    public SerialFilterFrame(CMRISystemConnectionMemo memo) {
=======
    SerialFilterFrame curFrame;

    public SerialFilterFrame() {
>>>>>>> master
```
they are showing a conflict.

Files showing a conflict that git recognizes get a little red icon beside the file (and its parent folders or packages) in the Projects or Files window. If that icon is showing, NetBeans can help you resolve the conflict by right-clicking on the tab title or file and selecting "Git->Resolve Conflicts" from the context menu.

If NetBeans does not show the conflict, you will need to fix the files by hand so that they compile correctly.

CCatania01 commented:
Is not a bracketed conflict.  The error is "cannot find symbol".
Is there a javax.vecmath.Vector3d jar missing?

bobjacobsen commented:
Is this the computer that you back-leveled? If so, it may be the problem that's described in the [Debugging section of the Mac OS X install guide](http://jmri.org/install/MacOSX.shtml).

CCatania01 commented:
That was the problem.  Clean compile now.
Regression testing to follow...

CCatania01 commented:
Preliminary testing has errors generated when adding sensors or turnouts to defined CMRI nodes.

The sequence is as follows:
- Select Tables->Sensors
- Select C/MRI connection
- Click Add
- Upon entering the 4th character for the System Name, the warning
 WARN  - bit number not in range 1 - 999 in CMRI system name: C2S2000
is written to the log.
The log output is definitely coming from CMRISystemConnectionMemo.
Same behavior for adding a Turnout.

I think this is happening upstream from any of the node changes I created.
The ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0] occurs when opening the panel file.  Only sensors and turnouts are defined,no GUI.
Here's a snippet of the log...
2017-09-15 20:13:02,058 util.Log4JUtil                        INFO  - ****** JMRI log ******* [main]
2017-09-15 20:13:03,557 util.Log4JUtil                        INFO  - This log is appended to file: /Users/Development/Library/Preferences/JMRI/log/messages.log [main]
2017-09-15 20:13:03,558 util.Log4JUtil                        INFO  - This log is stored in file: /Users/Development/Library/Preferences/JMRI/log/session.log [main]
2017-09-15 20:13:03,574 apps.Apps                             INFO  - PanelPro version 4.9.4ish+Development+20170916T0312Z+R701cfe8 starts under Java 1.8.0_144 on Mac OS X x86_64 v10.9.5 at Fri Sep 15 20:13:03 PDT 2017 [main]
2017-09-15 20:13:06,287 apps.Apps                             INFO  - Starting with profile Dual_CMRI_Connections.3f0749d2 [main]
2017-09-15 20:13:06,776 node.NodeIdentity                     INFO  - Using jmri-D49A20066074-3f0749d2 as the JMRI Node identity [main]
2017-09-15 20:13:07,227 serialdriver.SerialDriverAdapter      INFO  - cu.usbserial-A101OD7U port opened at 28800 baud with DTR: true RTS: true DSR: true CTS: true  CD: true [main]
2017-09-15 20:13:07,307 serialdriver.SerialDriverAdapter      INFO  - cu.usbserial-A101O9WE port opened at 28800 baud with DTR: true RTS: true DSR: true CTS: true  CD: true [main]
2017-09-15 20:13:07,919 util.FileUtilSupport                  INFO  - File path program: is /Users/Development/mrcs-cmri/2017-github/JMRI/ [main]
2017-09-15 20:13:07,921 util.FileUtilSupport                  INFO  - File path preference: is /Users/Development/Library/Preferences/JMRI/Dual_CMRI_Connections/ [main]
2017-09-15 20:13:07,922 util.FileUtilSupport                  INFO  - File path profile: is /Users/Development/Library/Preferences/JMRI/Dual_CMRI_Connections/ [main]
2017-09-15 20:13:07,926 util.FileUtilSupport                  INFO  - File path settings: is /Users/Development/Library/Preferences/JMRI/ [main]
2017-09-15 20:13:07,927 util.FileUtilSupport                  INFO  - File path home: is /Users/Development/ [main]
2017-09-15 20:13:07,928 util.FileUtilSupport                  INFO  - File path scripts: is /Users/Development/mrcs-cmri/2017-github/JMRI/jython/ [main]
2017-09-15 20:13:52,973 audio.JoalAudioFactory                INFO  - Initialised JOAL using OpenAL: vendor - OpenAL Community version - 1.1 ALSOFT 1.15.1 [Listed Table Generation]
2017-09-15 20:14:07,949 cmri.CMRISystemConnectionMemo         WARN  - bit number not in range 1 - 999 in CMRI system name: C2S2000 [AWT-EventQueue-0]
2017-09-15 20:14:07,949 cmri.CMRISystemConnectionMemo         WARN  - bit number not in range 1 - 999 in CMRI system name: C2S2000 [AWT-EventQueue-0]
2017-09-15 20:14:07,949 cmri.CMRISystemConnectionMemo         WARN  - bit number not in range 1 - 999 in CMRI system name: C2S2000 [AWT-EventQueue-0]
2017-09-15 20:14:07,950 cmri.CMRISystemConnectionMemo         WARN  - bit number not in range 1 - 999 in CMRI system name: C2S2000 [AWT-EventQueue-0]
2017-09-15 20:14:29,791 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-15 20:14:29,792 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-15 20:14:29,793 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-15 20:14:29,793 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-15 20:14:29,794 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]
2017-09-15 20:14:29,794 cmri.CMRISystemConnectionMemo         ERROR - illegal character in number field of CMRI system name: C2S20001 [AWT-EventQueue-0]

silverailscolo commented:
The Out of Range message could be caused by the entry pattern check I added to the Add New Item pane. That calls the SystemConnectionMemo. What would be the expected high limit for C/MRI?

KenC57 commented:
Chuck,

 

Convention for sensors and turnout take the connection name part from the
pull down for the connection. So in your case is would be the 2nd CMRI
connection that supplies the 'C2' part of the name. The 'S' would come from
it knowing it is creating a sensor. If you are expecting to enter the whole
finished system name (C2S2001), that would be against the convention used
for all other systems. It sounds like it is trying to apply the bit part
without thinking about the node number. I presume you are trying to enter
2001??

 

Now if you are trying to use the convention of a board and bit (B I think
was the separator) then something like 2B1 would mean node 2, bit 1. That
should generate C2S2001 as the system name.

 

Seems like the parsing is kicking in too soon or not considering which part
is the node number.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

pabender commented:
> On Sep 16, 2017, at 1:26 AM, Egbert Broerse <notifications@github.com> wrote:
> 
> The Out of Range message could be caused by the entry pattern check I added to the Add New Item pane. That calls the SystemConnectionMemo. What would be the expected high limit for C/MRI?
> 

Part of the entry for C/MRI is a node number, so the maximum value is on the order of the maximum number of nodes times 1000.

I can't remember, does the entry pattern check handle the alternate patterns, where bits can be specified separately ( separated by colons, I believe. )?

Paul

CCatania01 commented:
The numeric limit for sensor and turnout is 999, although a SUSIC can have up to 1024 bits.  The alternate format deals with that extended range.

The System name is formed with node address multiplied by 1000 + bit number.

Ken:  I always use two digit node addresses in our client work.  The example was showing connection CMRI2, bit 1 for node 20.

Parsing IS kicking in too soon, and incorrectly.  I do not think predictive/auto fill/look ahead is needed for these data fields; range and format checking, yes.

The alternate format should work, although I have not gotten that far in the regression testing.

CCatania01 commented:
Who would be responsible for fixing CMRISystemConnectionMemo?
Reverting that class back to the 4.9.3ish version is an acceptable option.
I cannot move forward until this is corrected.

rhwood commented:
> Who would be responsible for fixing CMRISystemConnectionMemo?

You are responsible for making any necessary changes for this PR.

> Reverting that class back to the 4.9.3ish version is an acceptable option.
> I cannot move forward until this is corrected.

The only changes I see to that file between 4.9.2 and current JMRI/master in CMRISystemConnectionMemo are:
- Removal of unused imports
- Javadocs
- Some messages that were logged as errors now logged as warnings

Are you certain the problems you are seeing are in the CMRISystemConnectionMemo and not really elsewhere, and that the changes that are causing these problems are not only in your own fork (meaning I can't see them)?

bobjacobsen commented:
@CCatania01: I think you misunderstand a couple things.

First, "CS2000" was never a valid sensor name.  CS2001 is the 1st (ordinal) number on the 2nd node. "2000" generates this:
![skitch](https://user-images.githubusercontent.com/2774117/30522593-971e17b8-9b87-11e7-8ca4-1872ba656a65.png)

Second, you don't type the entire system name when doing a add via the name.  To do what you want (assuming 2001), you configure like this:
![skitch](https://user-images.githubusercontent.com/2774117/30522609-d586e7a0-9b87-11e7-867f-13c9fbb1f089.png)
Note that you select the system name from the drop down box, and add a _hardware_ address, then you click create. It will then show up in the sensor list with its correct system name:
![skitch](https://user-images.githubusercontent.com/2774117/30522626-25283480-9b88-11e7-8a1a-707fcd5b94d1.png)
(In this example, I have three C/MRI connections configured, but the idea is the same) 

The [window help for the "Add New Sensor" window](http://jmri.org/help/en/package/jmri/jmrit/beantable/SensorAddEdit.shtml) covers all this, including specific C/MRI examples.

So it appears there's actually no problem here. The code is properly rejecting incompatible input.

CCatania01 commented:
Try node 20, 21, 30, or 102.
Yes the 000 was a typo.

Chuck

On Sep 17, 2017, at 9:19 AM, Bob Jacobsen wrote:

> @CCatania01: I think you misunderstand a couple things.
> 
> First, "CS2000" was never a valid sensor name. CS2001 is the 1st (ordinal) number on the 2nd node. "2000" generates this:
> 
> 
> Second, you don't type the entire system name when doing a add via the name. To do what you want (assuming 2001), you configure like this:
> 
> Note that you select the system name from the drop down box, and add a hardware address, then you click create. It will then show up in the sensor list with its correct system name:
> 
> (In this example, I have three C/MRI connections configured, but the idea is the same)
> 
> The window help for the "Add New Sensor" window covers all this, including specific C/MRI examples.
> 
> So it appears there's actually no problem here. The code is properly rejecting incompatible input.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bobjacobsen commented:
> Try node 20, 21, 30, or 102.

![skitch](https://user-images.githubusercontent.com/2774117/30523295-feb13b60-9b93-11e7-98df-121ba23bfac7.png)

@CCatania01, I'm done with this.  This is not my code, not my problem, but in the interest of being helpful I've spent way too much time trying to deal with inconsistent and partial information.  Open an [issue](https://github.com/JMRI/JMRI/issues/new) with screen shots of what you entered and what you saw, or even better, write a [unit test](http://jmri.org/help/en/html/doc/Technical/JUnit.shtml#testSwingCode) that shows the failure. Perhaps somebody will choose to look into it.

CCatania01 commented:
Will do.  Thanks.
Chuck

On Sep 17, 2017, at 10:39 AM, Bob Jacobsen wrote:

> Try node 20, 21, 30, or 102.
> 
> 
> 
> @CCatania01, I'm done with this. This is not my code, not my problem, but in the interest of being helpful I've spent way too much time trying to deal with inconsistent and partial information. Open an issue with screen shots of what you entered and what you saw, or even better, write a unit test that shows the failure. Perhaps somebody will choose to look into it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

CCatania01 commented:
Closing.

CCatania01 commented:
Re-open to continue...

CCatania01 commented:
What is the proper way to resolve a conflict?
Do I remove the code inline which is bracketed with <<<  ### >>>?
It looks like call includes the connection memo, those lines are correct.

CCatania01 commented:
Sorry, should have been more clear.
The lines in conflict should be removed/edited in the Resolve conflicts window?

pabender commented:
@CCatania01,

I'm not sure how your IDE is presenting the conflicts to you (I do most of my work on the command line) but you want to remove the lines that are incorrect from the file, including the markers showing where the conflict starts and ends and separating the two choices.

Paul

KenC57 commented:
Chuck,

 

Yes the idea is the system bracketed the two segments of the code in
conflict. The human then deletes the parts that don't belong, including
those brackets.

 

A rather manual operation but when it sees it as something it can't figure
out, this works.

-ken c

silverailscolo commented:
See my tip at the top of this thread and link you our online help file, Chuck. Carefully choose what you want to keep, it might be a case where you mix a bit of both code blocks. In the end, you should always remove the _3_ lines of <<< >>>> <<<

CCatania01 commented:
I was no sure where the change(s) needed to be made.
Thanks to all who commented.

rhwood commented on java/src/jmri/jmrix/cmri/serial/serialmon/Bundle.java:
```diff
@@ -24,7 +24,7 @@
 
     @Nullable
-    private static final String name = "jmri.jmrix.cmri.serial.serialmon.SerialFilterBundle"; // NOI18N
+    private static final String name = "jmri.jmrix.cmri.serial.serialmon.SerialFilterBundle"; // c2
```
Please retain the NOI18N comments unless you want someone to localize the String. See http://jmri.org/help/en/html/doc/Technical/I8N.shtml for an explanation.

CCatania01 replied:
Changed and read doc.

rhwood commented on java/src/jmri/jmrix/cmri/serial/serialmon/SerialMonFrame.java:
```diff
@@ -4,20 +4,78 @@
+    protected long lastTicks = 0L;
+    protected static int _DLE    = 0x10;    
+
+    final javax.swing.JFileChooser logFileChooser = new JFileChooser(FileUtil.getHomePath()); //jmri.jmrit.XmlFile.userFileLocationDefault());
```
Why are you thinking a user wants to browse his home directory and not the previously defined directory of JMRI files?

CCatania01 replied:
Restored back to default.

rhwood commented on java/src/jmri/jmrix/cmri/serial/serialmon/SerialMonFrame.java:
```diff
@@ -4,20 +4,78 @@
+        p.setSimplePreferenceState(alwaysOnTopCheck, alwaysOnTopCheckBox.isSelected());
+        p.setSimplePreferenceState(autoScrollCheck, !autoScrollCheckBox.isSelected());
+        _memo.getTrafficController().removeSerialListener(this);
+    }
```
You must *always* call super.dispose() in a dispose method to ensure that all resources in the frame are disposed.

CCatania01 replied:
Initially super.dispose() was called.  When exiting PanelPro an uncaught exception was thrown.
[Uncaught Exception Log.txt](https://github.com/JMRI/JMRI/files/1326613/Uncaught.Exception.Log.txt)
Without the call, the app exits cleanly.

initComponents in SerialMonFrame has additional items which are not in the Abstract class.  @Override is done on initComponents.
What is the correct approach for cleaning up in this instance?

rhwood commented on java/src/jmri/jmrix/cmri/serial/serialmon/SerialMonFrame.java:
```diff
@@ -31,13 +89,466 @@ protected void init() {
+    
+     /**
+     * Toggle the display on/off
+     * @param e 
```
All parameters in Javadoc comments need a description—Jenkins will list param annotations without descriptions as errors in the JavaDocs.

CCatania01 replied:
Removed @params

rhwood commented on java/test/jmri/jmrix/cmri/serial/cmrimetrics/CMRInetMetricsCollectorTest.java:
```diff
@@ -0,0 +1,45 @@
+    @After
+    public void tearDown() {
+        JUnitUtil.resetInstanceManager();
+        Log4JFixture.tearDown();
```
Please replace Log4JFixture.tearDown() with JUnitUtil.tearDown() and remove the previous line.

CCatania01 replied:
Replaced and removed line.

rhwood commented on java/test/jmri/jmrix/cmri/serial/cmrimetrics/CMRInetMetricsCollectorTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Before
+    public void setUp() {
+        Log4JFixture.setUp();
```
Please replace Log4JFixture.setUp() with JUnitUtil.setUp() and remove the following line.

CCatania01 replied:
Replaced and removed line.

rhwood commented on java/test/jmri/jmrix/cmri/serial/cmrimetrics/PackageTest.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class PackageTest{
+
+    // Main entry point
```
Please remove this main method. ant and  runtest.csh can run this package test (any test class even) without requiring this main method (this suppresses even normal success output from ant).

CCatania01 replied:
Do I completely remove the file?  Other test files in the cmri test class have this method.  I followed a common model for making the tests.

CCatania01 replied:
The test class was removed.  The main class is not implemented (yet) in this revision.

rhwood replied:
The JUnit 4-based test suites (like this one) need no body to the class, so you want everything before the class declaration and the class to just be:

```
public class PackageTest {}
```

pabender replied:
Definitely don't remove the packagetest file.

You MUST have at least one test in the package for this to pass CI tests.

rhwood replied:
Sorry that was probably not clear enough. You want something like this (plus any comments):

```
package jmri.jmrix.cmri.serial.cmrimetrics;

import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@RunWith(Suite.class)
@Suite.SuiteClasses({
    CMRInetMetricsCollectorTest.class,
    CMRInetMetricsDataTest.class
})

/**
 * Tests for the jmri.jmrix.cmri.serial.cmrimetrics package
 *
 * @author Chuck Catania Copyright (C) 2017
 */
public class PackageTest {}
```

rhwood commented on monitorLog.txt:
```diff
@@ -0,0 +1,30 @@
+Poll ua=30
+Poll ua=30
```
Did you intend to include this file in git?

CCatania01 replied:
Deleted file.

rhwood commented:
> +    @Override
> +    public void dispose() { 
> +        p.setSimplePreferenceState(timeStampCheck, timeCheckBox.isSelected());
> +        p.setSimplePreferenceState(rawDataCheck, rawCheckBox.isSelected());
> +        p.setSimplePreferenceState(alwaysOnTopCheck, alwaysOnTopCheckBox.isSelected());
> +        p.setSimplePreferenceState(autoScrollCheck, !autoScrollCheckBox.isSelected());
> +        _memo.getTrafficController().removeSerialListener(this);
> +    }
> Initially super.dispose() was called. When exiting PanelPro an uncaught exception was thrown.
> Uncaught Exception Log.txt <https://github.com/JMRI/JMRI/files/1326613/Uncaught.Exception.Log.txt>
> Without the call, the app exits cleanly.
> 
> initComponents in SerialMonFrame has additional items which are not in the Abstract class. @override <https://github.com/override> is done on initComponents.
> What is the correct approach for cleaning up in this instance?
> 
The correct thing would be remove the serial listener and then call super.dispose(). If that is still throwing errors, figure out what you have already nulled, and protect against access to it.

The problem with not calling super.dispose() is that the window itself never gets disposed of, which means that it never gets cleaned up by Java itself.

bobjacobsen commented:
Travis headless CI Findbugs found an issue:

```
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[INFO] Synchronization on SerialMonFrame.logStream in futile attempt to guard it [jmri.jmrix.cmri.serial.serialmon.SerialMonFrame] At SerialMonFrame.java:[line 491] ML_SYNC_ON_FIELD_TO_GUARD_CHANGING_THAT_FIELD
[INFO] 
```

CCatania01 commented:
Dispose is called at line 305 in the doneButton action listener.  The dispose method is in lines 73-79.  In the abstract class, (super)dispose saved checkbox states in SimplePreferences().  
Calling super.dispose accessed objects which were not active in the main class and threw an NPE.
Additional items were added to the frame.  The cleanup is done in the dispose() within the class with objects defined in the main class.

If this is not the proper way to do this, I will change it to the correct sequence.  Need advice on what that would be.

rhwood commented:
The right way to handle the dispose is to:
- Always call super.dispose() as the last call in any dispose() method (if any super class defines a dispose() method).
- Not duplicate any calls in super.dispose() within a dispose method.
- Not mask any fields in the super class (in this case AbstractMonFrame) in your sub-class (in this case SerialMonFrame) (if you need to change the value of any fields defined in the super class from the super class's definition, do so only in the sub-classes constructor or other methods called after construction).

You use NetBeans, so go to Preferences->Editor->Hints, select the Java language, and ensure every warning under "General" is checked except for "Use Functional Operations" (you can leave that one checked if you want to learn to use functional operations, but its not needed).

You will see warnings as light bulbs in the left of the editor and as little yellow lines to the right of the editor. Some of these warnings will indicate that you are masking a field defined in a super class, which can lead to subtle and hard to find errors, and you will want to unmask those fields by either deleting the field in the sub class (recommended unless it leads to compilation errors) or changing the name of the field in the sub class to not hide the super class's field.

bobjacobsen commented:
> On Sep 27, 2017, at 2:23 PM, Chuck Catania <notifications@github.com> wrote:
> 
> Dispose is called at line 305 in the doneButton action listener. The dispose method is in lines 73-79. In the abstract class, (super)dispose saved checkbox states in SimplePreferences().

Lots of things operate on Java Frames.  You don’t get to control that.  So dispose() should do super.dispose(). It should _not_ be done anywhere else. In particular, you shouldn’t be assuming that the doneButton action listener has complete control over the window.

> Calling super.dispose accessed objects which were not active in the main class and threw an NPE.

Then that should be fixed.

> Additional items were added to the frame. The cleanup is done in the dispose() within the class with objects defined in the main class.

I don’t know what distinction you’re making here between “within the class” and “objects defined in the main class”.  That seems to be a distinction to you, but they sound like the same thing to me.

CCatania01 commented:
By fixing the super.dispose in AbstractMonFrame, do you mean to remove the p.setSimplePreferencesState() calls in AbstractMonFrame?  The exception points to the first one.
As a test, I commented those out, called super.dispose() in SerialMonFrame and did not get the exception. I assumed I could not change the abstract class.

bobjacobsen commented:
> By fixing the super.dispose in AbstractMonFrame, do you mean to remove the p.setSimplePreferencesState() calls in AbstractMonFrame?

No, I very much don't mean that.  

`JmriJFrame` keeps track of position and size as preferences.  It can do that because there are various calls, including `dispose()`, that are invoked up through the implementations in `JmriJFrame`.  That code has worked for a long time, in multiple places, including in this class before your changes:
```
    @Override
    public void dispose() {
        _memo.getTrafficController().removeSerialListener(this);
        super.dispose();
    }
```
So the next step is to figure out why the changes are causing a problem when the window closes.  I think there's a clue above, where you talk about calling `dispose()` in the "doneButton action listener". It sounds like you're trying to manage the frame (window) state locally, instead of letting Java do it for you.  Is there a reason you need to do that?

bobjacobsen commented:
Just took a look at the changes.

You've reimplemented a lot of code from the `AbsMonFrame` super class within your `SerialMonFrame` class.  Is there a reason you have to do this?  One of the (many) principles behind object-oriented programing is "DRY":  Don't Repeat Yourself.  Having code in one place, not multiple ones, makes it easier to update in the future:  Bugs can be fixed in one place, features added in on place, and that work is available across the entire set of uses.

Here, that duplication is probably also the problem: The super class is still doing its work, still adding listeners to events, which are now duplicated by the ones in the `SerialMonFrame` class:  Things are being done twice.  And when those things are clean up, that clean up being done twice can cause problems.

CCatania01 commented:
Some background...
I wanted to update the SerialMonFrame GUI to have:
- A single button for pause/resume of the scrolled output.
- A single button for starting/stopping log output.
- An added checkbox for time difference display for packet transmissions.
- An added button to open the packet filter window.
- Some reformatting of the message display.
These additional items were added to the existing serial monitor window.
Additionally, I wanted to convert the processing by message type from a compound if-then-else to case statements.

I think at this point, it would be best to restore the original SerialMon GUI as there is really no value add to changing the user experience at this point.  The case statement code would remain in place.

I will explore how best to add the time diff checkbox and packet filter button .  I could use some advice on doing this.

silverailscolo commented:
From your description, it sounds like you suggest a user interface improvement that all monitors could use. Why not propose it as a change to that super class? Even more DRY

CCatania01 commented:
I have restored SerialMonFrame back to the original GUI.  I added a Filter Packets button and timeDiff checkbox in the subclass.  Filter Packets brings up a window for selecting node and packet type filtering.  The timeDiff checkbox is disabled as the code to insert the deltaT ms is not yet implemented.

Dispose() calls super.dispose and removes the serial listener.

CCatania01 commented:
Do I request a merge at this point?
My first PR the merge and close was done by someone else.

rhwood commented on java/src/jmri/jmrix/cmri/serial/serialmon/SerialMonFrame.java:
```diff
@@ -28,80 +58,101 @@ protected String title() {
+	try {
+		f.initComponents();
+            }
+	catch (Exception ex) {
```
Catch only the specific exceptions expected to be thrown here. See http://jmri.org/help/en/html/doc/Technical/RP.shtml on how to catch exceptions. See also http://jmri.org/help/en/html/doc/Technical/Logging.shtml for how to log them.

CCatania01 replied:
This form is used in many other classes.

CCatania01 commented:
Do I  request this PR be merged or is it picked up by the nightly build process?

CCatania01 commented:
Please merge #4035.

CCatania01 commented:
Why is the WIP label still attached?
What is preventing the merge?

pabender commented:
@bobjacobsen do you still have any concerns about this PR?

bobjacobsen commented:
> do you still have any concerns about this PR?

No concerns. Thanks to all involved.

CCatania01 commented:
Thanks to all who assisted and advised me on this PR.
Chuck Catania

On Oct 5, 2017, at 8:36 AM, Paul Bender wrote:

> Merged #4035.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
